extensions/websocket/02_websocketstream.js

```diff
@@ -0,0 +1,302 @@
+    #reject;
+    /** @type {(value: T | PromiseLike<T>) => void} */
+    #resolve;
+    /** @type {"pending" | "fulfilled"} */
+    #state = "pending";
+
+    constructor() {
+      this.#promise = new Promise((resolve, reject) => {
+        this.#resolve = resolve;
+        this.#reject = reject;
+      });
+    }
+
+    /** @returns {Promise<T>} */
+    get promise() {
+      return this.#promise;
+    }
+
+    /** @returns {"pending" | "fulfilled"} */
+    get state() {
+      return this.#state;
+    }
+
+    /** @param {any=} reason */
+    reject(reason) {
+      // already settled promises are a no-op
+      if (this.#state !== "pending") {
+        return;
+      }
+      this.#state = "fulfilled";
+      this.#reject(reason);
+    }
+
+    /** @param {T | PromiseLike<T>} value */
+    resolve(value) {
+      // already settled promises are a no-op
+      if (this.#state !== "pending") {
+        return;
+      }
+      this.#state = "fulfilled";
+      this.#resolve(value);
+    }
+  }
+
+  /**
+   * Tries to close the resource (and ignores BadResource errors).
+   * @param {number} rid
+   */
+  function tryClose(rid) {
+    try {
+      core.close(rid);
+    } catch (err) {
+      // Ignore error if the socket has already been closed.
+      if (!(err instanceof Deno.errors.BadResource)) throw err;
+    }
+  }
+
+  const _rid = Symbol("[[rid]]");
+  const _url = Symbol("[[url]]");
+  const _connection = Symbol("[[connection]]");
+  const _closed = Symbol("[[_closed]]");
+  class WebSocketStream {
+    [_rid];
+
+    [_url];
+    get url() {
+      return this[_url];
+    }
+
+    constructor(url, options) {
+      const prefix = "Failed to construct 'WebSocketStream'";
+      webidl.requiredArguments(arguments.length, 1, { prefix });
+      url = webidl.converters.USVString(url, {
+        prefix,
+        context: "Argument 1",
+      });
+      options = webidl.converters.WebSocketStreamOptions(options, {
+        prefix,
+        context: "Argument 1",
+      });
+
+      const wsURL = new URL(url);
+
+      if (wsURL.protocol !== "ws:" && wsURL.protocol !== "wss:") {
+        throw new DOMException(
+          "Only ws & wss schemes are allowed in a WebSocket URL.",
+          "SyntaxError",
+        );
+      }
+
+      if (wsURL.hash !== "" || wsURL.href.endsWith("#")) {
+        throw new DOMException(
+          "Fragments are not allowed in a WebSocket URL.",
+          "SyntaxError",
+        );
+      }
+
+      this[_url] = wsURL.href;
+
+      core.opSync("op_ws_check_permission", this[_url]);
+
+      if (
+        options.protocols.some((x) =>
+          options.protocols.indexOf(x) !== options.protocols.lastIndexOf(x)
+        )
+      ) {
+        throw new DOMException(
+          "Can't supply multiple times the same protocol.",
+          "SyntaxError",
+        );
+      }
+
+      core.opAsync("op_ws_create", {
+        url: this[_url],
+        protocols: options.protocols.join(", "),
```

Comment:
I realize `01_websocket.js` does this too but is this susceptible to header injection attacks with inputs like `{ protocols: ["foo", "\r\nx-pwned: true"] }`?

Reply:
I assume hyper/tungsenite sanitizes this. Should probably test that though.

Reply:
from looking at the `http` crate, this is the code that is used to check every byte in header values:
```rust
#[inline]
fn is_valid(b: u8) -> bool {
    b >= 32 && b != 127 || b == b'\t'
}
```
so we should be fine